File: Java/Arrays_que/Medium/nextPermutation.java
import java.util.*;

public class nextPermutation {
    static List<Integer> generateNextPermutation(List<Integer> li){
        int n = li.size();
        int ind = -1;

        for (int i = n-2; i >= 0; i--) {
            if(li.get(i) < li.get(i+1)){
                ind = i;
                break;
            }
        }

        //permutation is last permutation -> we will reverse it to get the first permutation
        if(ind == -1){
            Collections.reverse(li);
            return li;
        }

        for (int i = n-1; i > ind; i--) {
            if(li.get(i) > li.get(ind)){
                //swapping
                int temp = li.get(ind);
                li.set(ind, li.get(i));
                li.set(i, temp);
                break;
            }
        }

        List<Integer> tempLi = li.subList(ind+1, n);
        Collections.reverse(tempLi);
        return li;
    }
    public static void main(String[] args) {
        List<Integer> li = Arrays.asList(new Integer[] {2, 1, 5, 4, 3, 0, 0});
        List<Integer> ans = generateNextPermutation(li);

        System.out.println(ans);
    }
}

File: C++/Arrays/Medium/nextPermutation.cpp
#include <bits/stdc++.h>
#include <vector>
using namespace std;


class Solution {
public:
    void swap(vector<int>& nums, int i, int to){
        int temp = nums[i];
        nums[i] = nums[to];
        nums[to] = temp;
    }
    void reverse(vector<int> &nums, int from, int to){
        for(int i=from; i<=to; i++){
            swap(nums, i, to);
            to--;
        }
    } 
    void nextPermutation(vector<int>& nums) {
        // next_permutation(nums.begin(), nums.end());
        
        int n = nums.size();
        int ind = -1;

        for(int i=n-2; i>=0; i--){
            if(nums[i] < nums[i+1]){
                ind = i;
                break;
            }
        }

        if(ind != -1){
            for(int i=n-1; i>ind; i--){
                if(nums[i] > nums[ind]){
                    cout<<"Swap("<<nums[ind]<<" , "<<nums[i]<<")"<<endl;
                    swap(nums, i, ind);
                    break;
                }
            }
        }
        
        reverse(nums, ind+1, n-1);
    }
};

int main(){
    Solution sol;
    vector<int> nums = {2, 1, 5, 4, 3, 0, 0};
    sol.nextPermutation(nums);

    for(int i:nums){
        cout<<i<<" ";
    }
}
